target/linux/ath79/dts/qca9558_buffalo_wzr-450hp2.dts

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "qca955x.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "buffalo,wzr-450hp2", "qca,qca9558";
+	model = "Buffalo WZR-450HP2/WZR-450HP2D";
+
+	aliases {
+		led-boot = &led_system;
+		led-failsafe = &led_system;
+		led-running = &led_system;
+		led-upgrade = &led_system;
+		label-mac-device = &eth0;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_system: system {
+			label = "green:system";
+			gpios = <&gpio 20 GPIO_ACTIVE_LOW>;
+			default-state = "on";
+		};
+
+		wlan {
+			label = "green:wlan";
+			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy0tpt";
+		};
+
+		wps {
+			label = "green:wps";
+			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "Reset button";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+
+		rfkill {
+		       label = "RFKILL button";
+		       linux,code = <KEY_RFKILL>;
+		       gpios = <&gpio 21 GPIO_ACTIVE_LOW>;
+		       debounce-interval = <60>;
+		};
+	};
+};
+
+&spi {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x0 0x40000>;
+				read-only;
+			};
+
+			partition@40000 {
+				reg = <0x40000 0x10000>;
+				label = "u-boot-env";
+			};
+
+			partition@50000 {
+				compatible = "denx,uimage";
+				reg = <0x50000 0xf90000>;
+				label = "firmware";
+			};
+
+			partition@fe0000 {
+				reg = <0xfe0000 0x10000>;
+				label = "user_property";
+				read-only;
+			};
+
+			art: partition@ff0000 {
+				label = "art";
+				reg = <0xff0000 0x10000>;
+				read-only;
+
+				nvmem-layout {
+					compatible = "fixed-layout";
+					#address-cells = <1>;
+					#size-cells = <1>;
+
+					macaddr_wan: macaddr@0 {
+						compatible = "mac-base";
+						reg = <0x0 0x6>;
+						#nvmem-cell-cells = <1>;
+					};
+
+					macaddr_lan: macaddr@6 {
+						compatible = "mac-base";
+						reg = <0x6 0x6>;
+						#nvmem-cell-cells = <1>;
+					};
+
+					cal_ath9k: cal_ath9k@1000 {
+						reg = <0x1000 0x440>;
+					};
+				};
+			};
+		};
+	};
+};
+
+&mdio0 {
+	status = "okay";
+
+	phy0: ethernet-phy@0 {
```

Comment:
Wonder if this is appropriate to port to DSA.

Reply:
The eth0 cannot communicate with the switch ar8337 now, but works fine in 19.07. Comparing the [mach file](https://github.com/openwrt/openwrt/blob/a1837135e04b511a6204937c8ed00ba637cc6f0f/target/linux/ar71xx/files/arch/mips/ath79/mach-wzr-450hp2.c) and the current device tree, is there anything I missed to make eth0 and the corresponding switch port work properly?

Reply:
bad switch definition probably

```
######################START##############################
Found qca8k switch in ../openwrt/target/linux/ath79/dts/qca9558_buffalo_wzr-450hp2.dts
Conf for cpu port0, Mode sgmii, With qca,sgmii-rxclk-falling-edge
Conf for cpu port6, Mode rgmii-id
port1 led 0 is set to mode rule
port1 led 1 is set to mode rule
port1 led 2 is set to mode rule
port2 led 0 is set to mode rule
port2 led 1 is set to mode rule
port2 led 2 is set to mode rule
port3 led 0 is set to mode rule
port3 led 1 is set to mode rule
port3 led 2 is set to mode rule
#######################END###############################
```

I'm seeing a bunch of delays in ar71xx that are not present here.

At this point you should try DSA with qca8k.

Reply:
https://github.com/neheb/openwrt/commit/bdd73bb00a13e364c44db675780dd51d2fab5961 to get you started.

Reply:
You might also need to add

tx-internal-delay-ps
rx-internal-delay-ps

wherever appropriate.

Also because of some kind of bug in qca8k, you'll need to use external mdio (basically, move the mdio definitions outside of the switch one under mdio0).

Reply:
I managed to dump the registers from ar71xx by building an 19.07 image with [this trick](https://openwrt.org/docs/techref/hardware/switch/atheros). Now the eth0 works fine.

Reply:
Manually swapping the pad register values of port 0 and 6 with AR8337_PAD_MAC06_EXCHANGE_EN disabled also work, as in [DSA implementation](https://github.com/openwrt/openwrt/pull/17249#discussion_r1883179568)